### PR TITLE
feat: explicit routing headers codegen

### PR DESCRIPTION
**Pull request by viacheslav-rostovtsev**
this cl broadly includes two parts:
- updates to the ResoucePattern class and its segments, including:
   * converting patterns to regexes 
   * support for wildcard segments
   * separate segment type for the multivariate resource id segments  
- support for codegeneration for explicit routing headers, along with a new proto test

Won't build until Gax and CommonProtos changes are released (https://github.com/googleapis/gax-dotnet/pull/517)

### CODE REVIEW

File: Google.Api.Generator.Tests/ProtoTest.cs [hunk 1]
```diff
@@ -33,8 +33,10 @@ public class ProtoTest
             var protoPaths = protoFilenames.Select(x => Path.Combine(Invoker.GeneratorTestsDir, x));
             using (var desc = Invoker.TempFile())
             {
-                Invoker.Protoc($"-o {desc} --experimental_allow_proto3_optional --include_imports --include_source_info " +
```
Comment (jskeet) [inline Google.Api.Generator.Tests/ProtoTest.cs:36]: Revert this change? (It doesn't seem to do anything.) If we don't revert it, please outdent the second line of the `protocArgs` declaration.

File: Google.Api.Generator.Tests/ProtoTests/RoutingHeadersExplicit/RoutingHeadersExplicit.proto [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+package testing.routingheadersexplicit;
+
+option csharp_namespace = "Testing.RoutingHeadersExplicit";
+
+import "google/api/annotations.proto";
+import "google/api/client.proto";
+import "google/api/routing.proto";
+
+service RoutingHeadersExplicit {
+	option (google.api.default_host) = "routingheaders.example.com";
```
Comment (jskeet) [inline Google.Api.Generator.Tests/ProtoTests/RoutingHeadersExplicit/RoutingHeadersExplicit.proto:12]: I believe all our other proto files use spaces instead of tabs (and two-space indent).

File: Google.Api.Generator.Tests/ProtoTests/RoutingHeadersExplicit/RoutingHeadersExplicit.proto [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+	rpc Nested(NestedRequest) returns(Response) {
+		option (google.api.http) = {
+			get: "/v2/{nest1.nest2.name=projects/*/instances/*/tables/*}:withSubMessage"
+		};
+		option (google.api.routing) = {
+			routing_parameters {
+				field: "nest1.nest2.name"
+				path_template: "{table_name=projects/*/instances/*/tables/*}"
+			}
+			routing_parameters {
```
Comment (jskeet) [inline Google.Api.Generator.Tests/ProtoTests/RoutingHeadersExplicit/RoutingHeadersExplicit.proto:60]: Is there a reason why this test has two parameters? (I'd stick to just one to test nesting, and use the complex one for multiple parameters, unless there's something I'm missing.)

File: Google.Api.Generator.Tests/ProtoTests/RoutingHeadersExplicit/Testing.RoutingHeadersExplicit/RoutingHeadersExplicitClient.g.cs [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+            OnConstruction(grpcClient, effectiveSettings, clientHelper);
+            // TEST_END
+            
+            // for the easier reading, while the CR is in draft
+            _callPlainNoTemplate = clientHelper.BuildApiCall<SimpleRequest, Response>(grpcClient.PlainNoTemplateAsync, grpcClient.PlainNoTemplate, effectiveSettings.PlainNoTemplateSettings)
+                .WithGoogleRequestParam("name", request => request.Name);
+
+            _callPlainFullField = clientHelper.BuildApiCall<SimpleRequest, Response>(grpcClient.PlainFullFieldAsync, grpcClient.PlainFullField, effectiveSettings.PlainFullFieldSettings)
+                .WithExtractedGoogleRequestParam("table_name",
+                    new str::Regex("^(?<table_name>projects/[^/]+/instances/[^/]+/tables/[^/]+)/?$"),
```
Comment (jskeet) [inline Google.Api.Generator.Tests/ProtoTests/RoutingHeadersExplicit/Testing.RoutingHeadersExplicit/RoutingHeadersExplicitClient.g.cs:92]: Is there a particular reason for naming the groups within the regex? (At the moment I don't think those group names are used, but it's possible that they should be...)

File: Google.Api.Generator.Tests/ProtoTests/RoutingHeadersExplicit/Testing.RoutingHeadersExplicit/RoutingHeadersExplicitClient.g.cs [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+                    request => request.Name);
+
+            _callPlainExtract = clientHelper.BuildApiCall<SimpleRequest, Response>(grpcClient.PlainExtractAsync, grpcClient.PlainExtract, effectiveSettings.PlainExtractSettings)
+                .WithExtractedGoogleRequestParam("table_name",
+                    new str::Regex("^projects/[^/]+/instances/[^/]+/(?<table_name>tables/[^/]+)/?$"),
```
Comment (jskeet) [inline Google.Api.Generator.Tests/ProtoTests/RoutingHeadersExplicit/Testing.RoutingHeadersExplicit/RoutingHeadersExplicitClient.g.cs:97]: This is where the group is important, I suspect. If the value of `request.Name` is `projects/xyz/instances/abc/tables/def` then we want to send `table_name=tables/def` in the header, yes? I *think* it would currently send the whole thing.

File: Google.Api.Generator/Generation/MethodDetails.cs [hunk 6]
```diff
@@ -269,11 +271,50 @@ public Signature(ServiceDetails svc, MessageDescriptor msg , string sig)
...
+            /// </summary>
+            public sealed class FieldExtraction
+            {
+                public IEnumerable<FieldDescriptor> Fields { get; set; }
+                public string RegexStr { get; set; }
+                public bool NoRegexMatchingNeeded { get; set; }
+            }
+
+            public static RoutingHeader CreateOneFieldNoRegex(string encodedName, IEnumerable<FieldDescriptor> fields) =>
+                new()
```
Comment (jskeet) [inline Google.Api.Generator/Generation/MethodDetails.cs:300]: Note: we're planning on upgrading to the .NET 6 SDK as part of the GAX upgrade, so we could do the same thing here and use features up to and including C# 10.

File: Google.Api.Generator/Generation/MethodDetails.cs [hunk 7]
```diff
@@ -269,11 +271,50 @@ public Signature(ServiceDetails svc, MessageDescriptor msg , string sig)
...
+                {
+                    EncodedName = encodedName,
+                    Extractions = new List<FieldExtraction> { new() { Fields = fields, NoRegexMatchingNeeded = true } }
+                };
+        }
+
+        /// <summary>
+        /// A flat representation of the routing header annotation
+        /// with the order reified.
```
Comment (jskeet) [inline Google.Api.Generator/Generation/MethodDetails.cs:309]: Why do we need the order reified? GroupBy maintains ordering within a group, so I don't think we need this.

**Review by jskeet — COMMENTED** [review]
I haven't reviewed the details of the ResourcePattern changes, but we can do that later.

Using regular expressions still seems like overkill to me, but you're closer to the subject matter... if you think it's easier to use regexes than to add functionality to GAX, I guess I can live with it.

The part about what gets extracted is a concern at the moment - either I've misunderstood the feature, or we're currently not extracting the right parts of the string field.

File: Google.Api.Generator/ProtoUtils/ResourcePattern.cs [hunk 8]
```diff
@@ -123,25 +152,101 @@ void Check(bool cond, string msg)
...
+            public override string PathTemplateString => _givenSegment;
+            public override string RegexString => $"({_effectivePattern.RegexString})";
+            public override string Expand(IEnumerable<string> parameters) =>
+                parameters.First() + string.Join("", Separators.Zip(parameters.Skip(1), (s, p) => $"{s}{p}"));
+            public override string ToString() => PathTemplateString;
+
+            internal bool EndsWithDoubleWildcardPattern => _effectivePattern.EndsWithDoubleWildcardPattern;
+        }
+
+        // When double wildcard is by itself it can be anything
```
[the remark this answers is not in the document]
Comment (viacheslav-rostovtsev) [reply Google.Api.Generator/ProtoUtils/ResourcePattern.cs]: I've added some documentation to that effect.

File: Google.Api.Generator.Tests/ProtoTests/RoutingHeadersExplicit/RoutingHeadersExplicitFakes.cs [hunk 9]
```diff
@@ -0,0 +1,70 @@
+﻿// Copyright 2019 Google Inc. All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (viacheslav-rostovtsev) [reply Google.Api.Generator.Tests/ProtoTests/RoutingHeadersExplicit/RoutingHeadersExplicitFakes.cs]: 👍

File: Google.Api.Generator.Tests/ProtoTests/RoutingHeadersExplicit/RoutingHeadersExplicitFakes.cs [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Grpc.Core;
+using System;
+
```
[the remark this answers is not in the document]
Comment (viacheslav-rostovtsev) [reply Google.Api.Generator.Tests/ProtoTests/RoutingHeadersExplicit/RoutingHeadersExplicitFakes.cs:17]: 👍

File: Google.Api.Generator.Tests/ResourcePatternTest.cs [hunk 11]
```diff
@@ -14,28 +14,91 @@
...
+        [Theory]
+        [InlineData("{table_name=projects/*/instances/*/**}", "^(projects/[^/]+/instances/[^/]+(?:/.*)?)$")]
+        [InlineData("{table_name=regions/*/zones/*/**}", "^(regions/[^/]+/zones/[^/]+(?:/.*)?)$")]
+        [InlineData("{routing_id=projects/*}/**", "^(projects/[^/]+)(?:/.*)?$")]
+        [InlineData("{routing_id=projects/*/instances/*}/**", "^(projects/[^/]+/instances/[^/]+)(?:/.*)?$")]
+        [InlineData("{project_id=projects/*}/instances/*/**", "^(projects/[^/]+)/instances/[^/]+(?:/.*)?$")]
+        [InlineData("projects/*/{instance_id=instances/*}/**", "^projects/[^/]+/(instances/[^/]+)(?:/.*)?$")]
+        [InlineData("{project_id=projects/*}/**", "^(projects/[^/]+)(?:/.*)?$")]
+        [InlineData("subs/{sub.sub_name}", "^subs/([^/]+)/?$")]
+        [InlineData("{legacy.routing_id=**}", "^(.*)$")]
```
[the remark this answers is not in the document]
Comment (viacheslav-rostovtsev) [reply Google.Api.Generator.Tests/ResourcePatternTest.cs]: Updated

File: Google.Api.Generator.Tests/ResourcePatternTest.cs [hunk 12]
```diff
@@ -14,28 +14,91 @@
...
+            var pat = new ResourcePattern(pattern);
             Assert.Equal(expanded, pat.Expand(expandArgs));
         }
 
+        [Theory]
+        [InlineData("as/bs/cs", "as/bs/cs")]
+        [InlineData("*", "[^/]+")]
+        [InlineData("{a=*}", "([^/]+)")]
+        [InlineData("**", ".*")]
+        [InlineData("{a=**}", "(.*)")]
```
[the remark this answers is not in the document]
Comment (viacheslav-rostovtsev) [reply Google.Api.Generator.Tests/ResourcePatternTest.cs]: The one without a resource id -- yes, the one with -- no.

File: Google.Api.Generator/Generation/MethodDetails.cs (hunk 6 above)
Comment (viacheslav-rostovtsev) [reply Google.Api.Generator/Generation/MethodDetails.cs]: Probably not worth it yet?

File: Google.Api.Generator.Tests/ProtoTests/RoutingHeadersExplicit/Testing.RoutingHeadersExplicit/RoutingHeadersExplicitClient.g.cs (hunk 5 above)
Comment (viacheslav-rostovtsev) [reply Google.Api.Generator.Tests/ProtoTests/RoutingHeadersExplicit/Testing.RoutingHeadersExplicit/RoutingHeadersExplicitClient.g.cs]: 👍

File: Google.Api.Generator.Tests/ProtoTests/RoutingHeadersExplicit/Testing.RoutingHeadersExplicit/RoutingHeadersExplicitClient.g.cs (hunk 4 above)
Comment (viacheslav-rostovtsev) [reply Google.Api.Generator.Tests/ProtoTests/RoutingHeadersExplicit/Testing.RoutingHeadersExplicit/RoutingHeadersExplicitClient.g.cs]: 👍

File: Google.Api.Generator.Tests/ProtoTests/RoutingHeadersExplicit/RoutingHeadersExplicit.proto (hunk 3 above)
Comment (viacheslav-rostovtsev) [reply Google.Api.Generator.Tests/ProtoTests/RoutingHeadersExplicit/RoutingHeadersExplicit.proto]: It served a purpose for the previous design, now it is refactored to ... still have two parameters but to showcase a different thing entirely.

**Comment by viacheslav-rostovtsev** [comment]
Please merge it if it's OK to do so with the beta gax referenced.

File: Google.Api.Generator/Generation/MethodDetails.cs [hunk 13]
```diff
@@ -476,28 +534,73 @@ private IEnumerable<RoutingHeader> ReadRoutingHeaders(HttpRule http, MessageDesc
...
+                if (parameterName == null)
+                {
+                    var msg = $"The explicit routing header pattern {param.PathTemplate} should have exactly one non-multivariate ResourceId segment";
+                    throw new InvalidOperationException(
+                        msg);
+                }
+
+                // The full pattern should match the entire field and allow for an
+                // optional `/` at the end.
+                var patternRegex = pattern.FullFieldRegexString;
```
[the remark this answers is not in the document]
Comment (jskeet) [reply Google.Api.Generator/Generation/MethodDetails.cs:566]: I think it would be fine to add that as a comment just above the specific property setter. Not a blocker, but something to consider next time we're in this file.

File: Google.Api.Generator/Generation/MethodDetails.cs [hunk 14]
```diff
@@ -476,28 +534,73 @@ private IEnumerable<RoutingHeader> ReadRoutingHeaders(HttpRule http, MessageDesc
...
+            }
+
+            List<FieldDescriptor> SplitVerifyFieldPath(string path, MessageDescriptor desc)
+            {
+                var fields = new List<FieldDescriptor>();
+                foreach (var fieldName in path.Split('.'))
+                {
+                    if (desc == null)
+                    {
+                        throw new InvalidOperationException($"Invalid path in http url: '{path}'. '{fields.Last().Name}' does not have subfields.");
```
[the remark this answers is not in the document]
Comment (jskeet) [reply Google.Api.Generator/Generation/MethodDetails.cs]: This still leaves the odd check here, which is basically broken. I'm happy to merge and then create another PR that removes it though.

File: Google.Api.Generator.Tests/ResourcePatternTest.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (jskeet) [reply Google.Api.Generator.Tests/ResourcePatternTest.cs]: I'm not seeing this one being updated. I saw the other comment about only some cases wanting `+` instead of `*` - I'll take a more careful look before merging.

File: Google.Api.Generator.Tests/ProtoTests/RoutingHeadersExplicit/RoutingHeadersExplicit.proto [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+      get: "/v2/{name=projects/*/instances/*/tables/*}:sampleRowKeys"
+    };
+    option (google.api.routing) = {
+      routing_parameters {
+        field: "name"
+      }
+    };
+  }
+
+  rpc PlainNoExtraction(SimpleRequest) returns(Response) {
```
Comment (jskeet) [inline Google.Api.Generator.Tests/ProtoTests/RoutingHeadersExplicit/RoutingHeadersExplicit.proto:27]: I missed this before - please fix the indentation. (Again, that could be post merge.)

File: Google.Api.Generator.Tests/ResourcePatternTest.cs [hunk 16]
```diff
@@ -21,21 +21,113 @@ namespace Google.Api.Generator.Tests
...
+        [InlineData("as/{a=**}/cs", "as/(.+)/cs")]
+        [InlineData("as/{a=**/cs}", "as/(.*/cs)")]
+        [InlineData("as/bs/{a=**}", "as/bs/(.+)")]
+        [InlineData("as/{a=bs/**}", "as/(bs(?:/.*)?)")]
+
+        [InlineData("as/*/bs/**", "as/[^/]+/bs(?:/.*)?")]
+        [InlineData("as/{a=*}/bs/{b=**}", "as/([^/]+)/bs/(.+)")]
+        [InlineData("{a=as/*}/{b=bs/**}", "(as/[^/]+)/(bs(?:/.*)?)")]
+        [InlineData("as/{a=*}/bs/{b=**}", "as/([^/]+)/bs/(.+)")]
+        [InlineData("{a=as/*}/{b=bs/**}", "(as/[^/]+)/(bs(?:/.*)?)")]
```
Comment (jskeet) [inline Google.Api.Generator.Tests/ResourcePatternTest.cs:94]: And this one is from line 92

File: Google.Api.Generator.Tests/ResourcePatternTest.cs [hunk 17]
```diff
@@ -21,21 +21,113 @@ namespace Google.Api.Generator.Tests
...
+        [InlineData("{a=**/bs}/cs", "(.*/bs)/cs")]
+        [InlineData("as/{a=**}/cs", "as/(.+)/cs")]
+        [InlineData("as/{a=**/cs}", "as/(.*/cs)")]
+        [InlineData("as/bs/{a=**}", "as/bs/(.+)")]
+        [InlineData("as/{a=bs/**}", "as/(bs(?:/.*)?)")]
+
+        [InlineData("as/*/bs/**", "as/[^/]+/bs(?:/.*)?")]
+        [InlineData("as/{a=*}/bs/{b=**}", "as/([^/]+)/bs/(.+)")]
+        [InlineData("{a=as/*}/{b=bs/**}", "(as/[^/]+)/(bs(?:/.*)?)")]
+        [InlineData("as/{a=*}/bs/{b=**}", "as/([^/]+)/bs/(.+)")]
```
Comment (jskeet) [inline Google.Api.Generator.Tests/ResourcePatternTest.cs:93]: Not sure why it's giving this error twice, but it's right - there's a duplicate on line 91.

**Review by jskeet — COMMENTED** [review]
Some more small things that *could* be fixed post-merge.

File: Google.Api.Generator.Tests/ProtoTests/RoutingHeadersExplicit/RoutingHeadersExplicit.proto (hunk 15 above)
Comment (viacheslav-rostovtsev) [reply Google.Api.Generator.Tests/ProtoTests/RoutingHeadersExplicit/RoutingHeadersExplicit.proto]: I was not sure what exactly was wrong with the indentation. Reformatted the file.

File: Google.Api.Generator.Tests/ResourcePatternTest.cs (hunk 17 above)
Comment (viacheslav-rostovtsev) [reply Google.Api.Generator.Tests/ResourcePatternTest.cs]: Removed the duplicates.

File: Google.Api.Generator.Tests/ResourcePatternTest.cs (hunk 16 above)
Comment (viacheslav-rostovtsev) [reply Google.Api.Generator.Tests/ResourcePatternTest.cs]: Removed the duplicates.

File: Google.Api.Generator/Generation/MethodDetails.cs [hunk 18]
```diff
@@ -297,20 +296,18 @@ public sealed class FieldExtraction
             }
 
             public static RoutingHeader CreateOneFieldNoRegex(string encodedName, IEnumerable<FieldDescriptor> fields) =>
-                new()
+                new RoutingHeader()
```
Comment (jskeet) [inline Google.Api.Generator/Generation/MethodDetails.cs]: Nit (for me to fix in a follow-up PR) - we normally omit the `()` when it's not needed. (Here and below for FieldExtraction.)

File: Google.Api.Generator/Generation/MethodDetails.cs (hunk 6 above)
Comment (jskeet) [reply Google.Api.Generator/Generation/MethodDetails.cs]: This was more of a "note for the future" than a proposed change now :)

**Review by jskeet — APPROVED** [review]
I'm going to squash and merge this now, then send a follow-up PR to fix a few nits. I believe there's another issue (around multiple header values) that we want to address before we do a release, but that'll be easier to check after merging.

File: Google.Api.Generator/Generation/MethodDetails.cs (hunk 14 above)
[the remark this answers is not in the document]
Comment (jskeet) [reply Google.Api.Generator/Generation/MethodDetails.cs]: Ah, that's now been fixed - I must have been looking at an outdated version.

File: Google.Api.Generator/Generation/MethodDetails.cs (hunk 18 above)
Comment (viacheslav-rostovtsev) [reply Google.Api.Generator/Generation/MethodDetails.cs]: :+1:
